spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java

```diff
@@ -0,0 +1,198 @@
+import org.springframework.dao.DuplicateKeyException;
+import org.springframework.data.mongodb.MongoDbFactory;
+import org.springframework.data.mongodb.core.CollectionCallback;
+import org.springframework.data.mongodb.core.MongoTemplate;
+import org.springframework.data.mongodb.core.query.Criteria;
+import org.springframework.data.mongodb.core.query.Query;
+import org.springframework.data.mongodb.core.query.Update;
+import org.springframework.integration.metadata.ConcurrentMetadataStore;
+import org.springframework.util.Assert;
+
+import com.mongodb.BasicDBObject;
+import com.mongodb.DBCollection;
+import com.mongodb.MongoException;
+
+/**
+ * MongoDbMetadataStore implementation of {@link ConcurrentMetadataStore}.
+ * Use this {@link org.springframework.integration.metadata.MetadataStore} to
+ * achieve meta-data persistence shared across application instances and
+ * restarts.
+ *
+ * @author Senthil Arumugam, Samiraj Panneer Selvam
+ * @since 4.2
+ *
+ */
+public class MongoDbMetadataStore implements ConcurrentMetadataStore {
+
+	protected final Log logger = LogFactory.getLog(getClass());
+	private final MongoTemplate template;
+	private static final String DEFAULT_COLLECTION_NAME = "metadataStore";
+	private final String collectionName;
+	private static final String ID_FIELD = "_id";
+	private static final String VALUE = "value";
+	
+	/**
+	 * 
+	 * Configure the MongoDbMetadataStore by provided {@link MongoTemplate} and
+	 * default collection name - {@link #DEFAULT_COLLECTION_NAME}.
+	 * 
+	 * @param template The mongodb template
+	 * @param collectionName the collection name where it persists the data
+	 */
+	public MongoDbMetadataStore(MongoTemplate template, String collectionName) {
+		Assert.notNull(template, "'template' must not be null.");
+		Assert.hasText(collectionName, "'collectionName' must not be empty.");
+		this.template = template;
+		this.collectionName = collectionName;
+	}
+
+	/**
+	 * Configure the MongoDbMetadataStore by provided {@link MongoTemplate} and
+	 * default collection name - {@link #DEFAULT_COLLECTION_NAME}.
+	 * 
+	 * @param template The mongodb template
+	 */
+	public MongoDbMetadataStore(MongoTemplate template) {
+		this(template, DEFAULT_COLLECTION_NAME);
+	}
+
+	/**
+	 * Configure the MongoDbMetadataStore by provided {@link MongoDbFactory} and
+	 * default collection name - {@link #DEFAULT_COLLECTION_NAME}.
+	 * 
+	 * @param factory The mongodb factory
+	 */
+	public MongoDbMetadataStore(MongoDbFactory factory) {
+		this(factory, DEFAULT_COLLECTION_NAME);
+	}
+	
+	/**
+	 * Configure the MongoDbMetadataStore by provided {@link MongoDbFactory} and
+	 * collection name
+	 * 
+	 * @param collectionName the collection name where it persists the data
+	 * @param factory MongoDbFactory
+	 */
+	public MongoDbMetadataStore(MongoDbFactory factory, String collectionName) {
+		this(new MongoTemplate(factory),collectionName);
+	}
+	
+	
+	/**
+	 * @param key String
+	 * @param value String
```

Comment:
Please, add the comprehensive description for params.
And I recommend to add method description to address which MongoDB operation is performed